# Exploratory Data Analysis of fuel_quality_data

In [42]:
# importing the required library

%matplotlib inline
import numpy as np
import pandas as pd
import seaborn as sns
from scipy.stats import kurtosis
import matplotlib.pyplot as plt
plt.rc('figure', figsize=(12, 12))
sns.set(color_codes=True)

In [19]:
# View the datasets to gain insight of the data

url2 = "https://raw.githubusercontent.com/WalePhenomenon/climate_change/master/fuel_ferc1.csv"

fuel_quality_data = pd.read_csv(url2, error_bad_lines=False)
fuel_quality_data.describe(include="all")

record_id  utility_id_ferc1   report_year  \
count                     29523      29523.000000  29523.000000   
unique                    29523               NaN           NaN   
top     f1_fuel_2013_12_41_1_13               NaN           NaN   
freq                          1               NaN           NaN   
mean                        NaN        118.601836   2005.806050   
std                         NaN         74.178353      7.025483   
min                         NaN          1.000000   1994.000000   
25%                         NaN         55.000000   2000.000000   
50%                         NaN        122.000000   2006.000000   
75%                         NaN        176.000000   2012.000000   
max                         NaN        514.000000   2018.000000   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
count             29523               29523     29343     2.952300e+04   
unique             2315                   6         9              NaN   
top           big stone                 gas       mcf              NaN   
freq                156               11486     11354              NaN   
mean                NaN                 NaN       NaN     2.622119e+06   
std                 NaN                 NaN       NaN     9.118004e+06   
min                 NaN                 NaN       NaN     1.000000e+00   
25%                 NaN                 NaN       NaN     1.381700e+04   
50%                 NaN                 NaN       NaN     2.533220e+05   
75%                 NaN                 NaN       NaN     1.424034e+06   
max                 NaN                 NaN       NaN     5.558942e+08   

        fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
count          29523.000000               29523.000000   
unique                  NaN                        NaN   
top                     NaN                        NaN   
freq                    NaN                        NaN   
mean               8.492111                 208.649031   
std               10.600220                2854.490090   
min                0.000001                -276.080000   
25%                1.024000                   5.207000   
50%                5.762694                  26.000000   
75%               17.006000                  47.113000   
max              341.260000              139358.000000   

        fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
count                   2.952300e+04         29523.000000  
unique                           NaN                  NaN  
top                              NaN                  NaN  
freq                             NaN                  NaN  
mean                    9.175704e+02            19.304354  
std                     6.877593e+04          2091.540939  
min                    -8.749370e+02           -41.501000  
25%                     3.778500e+00             1.940000  
50%                     1.737100e+01             4.127000  
75%                     4.213700e+01             7.745000  
max                     7.964521e+06        359278.000000

# Question 1

if you are given two lists:
A = [1, 2, 3, 4, 5, 6], B = [13, 21, 34]
The task is to create a list with the dimesion of A and B in a single dimenstion with output:
A_B = [1, 2, 3, 4, 5, 6, 13, 21, 34]
which of the following option is the best way to create this list?

##### Answer: None of the above

# Question 2

###### how to create an identity matrix

In [23]:
# answer

np.eye(3)
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

# Question 3

which of the folowing fuel type code has the lowest average fuel cost per unit burned?

In [25]:
fuel_quality_data.groupby("fuel_type_code_pudl")["fuel_cost_per_unit_burned"].mean().sort_values()[:2]

fuel_type_code_pudl
gas      13.659397
other    18.253856
Name: fuel_cost_per_unit_burned, dtype: float64

##### Answer: Gas

# Question 4

what is the standard deviation and the 75th percentile of the measure of energy per unit (Fuel_mmbtu_per_unit) in two decimal places?

In [28]:
std = fuel_quality_data["fuel_mmbtu_per_unit"].std()
percentile = np.percentile(fuel_quality_data["fuel_mmbtu_per_unit"], 75)

print(std)
print(percentile)

10.600220307806886
17.006


##### Answer: 10.60 and 17.01

# Question 5

What is the skewness and kurtosis for the fuel quality burned in two decimal places?

In [45]:
skew = fuel_quality_data["fuel_qty_burned"].skew()
kurt = kurtosis(fuel_quality_data["fuel_qty_burned"])

print(skew)
print(kurt)

15.851495469109503
651.2589365474387


##### Answer: 15.85 and  651.37

# Question 6

which of the features has missing values and what is the total number of missing value and percentage of the misssing rows as a factor of the total number of rows in three decimal places?

In [46]:
# checking for missing data

fuel_quality_data.isnull().sum() # we have 180 missing data in the fuel_unit column

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [48]:
180 / len(fuel_quality_data["fuel_unit"])

0.00609694136774718

##### Answer: Feature: fuel_unit, Total: 180, Percentage: 0.006

# Question 7

the feature with missing value falls under which category? what missing value imputation technique would you use?

In [49]:
# use groupby to count the sum of each unique value in the fuel_unit column

fuel_quality_data.groupby('fuel_unit')['fuel_unit'].count()
fuel_quality_data[["fuel_unit"]] = fuel_quality_data[["fuel_unit"]].fillna(value="mcf")

# the above code will replace the missing values in fuel_unit column with the most common unit which is the "mcf"

In [51]:
# checking if missing value have being filled 

fuel_quality_data.isnull().sum() # we are good to go now

record_id                       0
utility_id_ferc1                0
report_year                     0
plant_name_ferc1                0
fuel_type_code_pudl             0
fuel_unit                       0
fuel_qty_burned                 0
fuel_mmbtu_per_unit             0
fuel_cost_per_unit_burned       0
fuel_cost_per_unit_delivered    0
fuel_cost_per_mmbtu             0
dtype: int64

##### Answer: Categorical and mode imputation

# Question 8

which of the features has the second and third lowest correlation with the fuel cost per unit burned?

In [52]:
fuel_quality_data.corr().loc["fuel_cost_per_unit_burned"].sort_values(ascending=True)

utility_id_ferc1               -0.037863
fuel_qty_burned                -0.018535
fuel_mmbtu_per_unit            -0.010034
fuel_cost_per_mmbtu            -0.000437
fuel_cost_per_unit_delivered    0.011007
report_year                     0.013599
fuel_cost_per_unit_burned       1.000000
Name: fuel_cost_per_unit_burned, dtype: float64

##### Answer: fuel_qty_burned and fuel_mmbtu_per_unit

# Question 9

For the fuel type coal, what is the percentage change in the fuel cost per unit burned in 1998 compared to 1994?

In [53]:
yearly_cost = fuel_quality_data.groupby(["report_year", "fuel_type_code_pudl"])["fuel_cost_per_unit_burned"].sum()

year_1994, year_1998 = yearly_cost.loc[[1994, 1998], "coal"]
percentage_change = ((year_1998 - year_1994) / (year_1994)) * 100
percentage_change

-20.56765451826049

##### Answer: -21%

# Question 10

which year has the average fuel cost per unit delivered?

In [54]:
fuel_quality_data.groupby("report_year")["fuel_cost_per_unit_delivered"].mean().sort_values()

report_year
1999       25.551627
1995       32.735269
2006       38.657484
2005       41.438184
2007       43.325023
2017       46.196861
2002       47.594361
2003       55.663493
2008       58.588197
2011       59.774667
2001       60.050396
2012       60.994502
1994       63.636060
2010       91.862105
2016      103.901761
2004      139.524275
2013      172.307591
2014      192.737183
1998      287.154420
2015      326.535511
2018      499.269966
2009      652.694163
2000      985.362877
1996     9196.705948
1997    11140.197239
Name: fuel_cost_per_unit_delivered, dtype: float64

##### Answer: 1997